pkg/operator/ceph/object/zone/controller.go

```diff
@@ -0,0 +1,280 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/coreos/pkg/capnslog"
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	controllerName = "ceph-object-store-zone-controller"
+	waitTime       = 10 * time.Second
+)
+
+var WaitForRequeueIfObjectStoreZoneGroupNotReady = reconcile.Result{Requeue: true, RequeueAfter: waitTime}
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", controllerName)
+
+var cephObjectStoreZoneKind = reflect.TypeOf(cephv1.CephObjectStoreZone{}).Name()
+
+// Sets the type meta for the controller main object
+var controllerTypeMeta = metav1.TypeMeta{
+	Kind:       cephObjectStoreZoneKind,
+	APIVersion: fmt.Sprintf("%s/%s", cephv1.CustomResourceGroup, cephv1.Version),
+}
+
+// ReconcileObjectStoreZone reconciles a ObjectStoreZone object
+type ReconcileObjectStoreZone struct {
+	client  client.Client
+	scheme  *runtime.Scheme
+	context *clusterd.Context
+}
+
+// Add creates a new CephObjectStoreZone Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, context *clusterd.Context) error {
+	return add(mgr, newReconciler(mgr, context))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, context *clusterd.Context) reconcile.Reconciler {
+	// Add the cephv1 scheme to the manager scheme so that the controller knows about it
+	mgrScheme := mgr.GetScheme()
+	cephv1.AddToScheme(mgr.GetScheme())
+
+	return &ReconcileObjectStoreZone{
+		client:  mgr.GetClient(),
+		scheme:  mgrScheme,
+		context: context,
+	}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes on the CephObjectStoreZone CRD object
+	err = c.Watch(&source.Kind{Type: &cephv1.CephObjectStoreZone{TypeMeta: controllerTypeMeta}}, &handler.EnqueueRequestForObject{}, opcontroller.WatchControllerPredicate())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile reads that state of the cluster for a CephObjectStoreZone object and makes changes based on the state read
+// and what is in the CephObjectStoreZone.Spec
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileObjectStoreZone) Reconcile(request reconcile.Request) (reconcile.Result, error) {
```

Comment:
Per our discussion, I see now that the controllers are already creating the resources as expected, so you can ignore my comment. The takeaway from the discussion is that an improvement for a near future PR is to commit the period more often, rather than just with the object store creation. For example, if a non-master zonegroup is created, its controller could commit the period instead of waiting for an object store to commit it.